python/mlx/data/datasets/speechcommands.py

```diff
@@ -0,0 +1,143 @@
+                if h.hexdigest() != target_hash:
+                    raise RuntimeError(
+                        f"[speechcommands] File download corrupted. sha256sums don't match. Please manually delete {str(target_compressed)}."
+                    )
+        gzip_decompress(target_compressed, target, quiet=quiet)
+        target_compressed.unlink()
+    return target
+
+
+def get_metadata(tarfile_path):
+    """
+    Helper function to get metadata from the tarfile.
+    This includes file names by split, as well as class names and their mapping to integers.
+    """
+    output = {}
+    with tarfile.open(tarfile_path) as tar:
+        for split, split_file in SPLITS_INFO.items():
+            if split_file is None:
+                continue
+            f = tar.extractfile(split_file)
+            fileslist = ["./" + s.decode().strip() for s in f.readlines()]
+            output[split] = fileslist
+
+        all_wav_files = [
+            f.name
+            for f in tar.getmembers()
+            if ".wav" in f.name
+            and HASH_DIVIDER in f.name
+            and EXCLUDE_FOLDER not in f.name
+        ]
+    output["train"] = set(all_wav_files) - set(output["validation"] + output["test"])
+    classes = dict(
+        map(reversed, enumerate(sorted(set(f.split("/")[-2] for f in output["train"]))))
+    )
+    return output, classes
+
+
+def _to_audio(sample):
+    filename = bytes(sample["file"]).decode()
+    label = bytes(filename.split("/")[-2].encode())
+    return {"file": sample["file"], "audio": sample["audio"], "label": label}
+
+
+def load_speechcommands(
+    root=None,
+    split="train",
+    quiet=False,
+    validate_download=True,
+    prefetch_size=8,
+    num_threads=4,
+):
+    """Load the Speech Commands (v0.0.2) [1] dataset directly from the TAR archive.
+
+    Args:
+        root (Path or str, optional): The The directory to load/save the data. If
+            none is given the ``~/.cache/mlx.data/speechcommands`` is used.
+        split (str): The split to use. It should be one of train,
+            validation or test
+        quiet (bool): If true do not show download (and possibly decompression)
+            progress.
+        prefetch_size (int, optional): The number of samples for prefetching. Default is 8.
+        num_threads (int, optional): The number of threads to use for prefetching. Default is 4.
+
+
+    References
+    ----------
+    Warden, Pete. "Speech commands: A dataset for limited-vocabulary speech recognition." arXiv preprint arXiv:1804.03209 (2018).
+    """
+    target = download_speechcommands(
+        root=root, quiet=quiet, validate_download=validate_download
+    )
+    target = str(target)
+    assert (
+        split in SPLITS_INFO
+    ), f"Unknown split {split}. Should be one of [{', '.join(SPLITS_INFO.keys())}]"
+
+    files_by_split, class_map = get_metadata(target)
+
+    dset = (
+        dx.files_from_tar(target)
```

Comment:
Since you parsed the files above you can change this to a buffer directly no need to have a stream and then a buffer. For instance you could have 

```
file_list = [
    {"file": f.encode(), "label": class_map[f.split("/")[-2]]} for f in files_by_split[split]
]
dset = (
    dx.buffer_from_vector(file_list)
    .read_from_tar(...)
    .load_audio(...)
)
```

This would remove the need for `_to_audio` as well as prefetching etc.

Reply:
Oh great. That's very neat! Updated and pushed!